File: tests/bench/io_mmap_throughput.c
/**
 * io_mmap_throughput.c — I/O 基线：大文件 mmap 顺序扫描（C -O2 对照）
 * 路径须与 run-perf-io.sh 生成的 tests/bench/.io_mmap_bench_tmp 一致。
 */
#include <fcntl.h>
#include <stdint.h>
#include <stdio.h>
#include <sys/mman.h>
#include <sys/stat.h>
#include <unistd.h>

int main(void) {
  const char *path = "tests/bench/.io_mmap_bench_tmp";
  int fd = open(path, O_RDONLY);
  if (fd < 0) {
    perror("open");
    return 1;
  }
  struct stat st;
  if (fstat(fd, &st) != 0) {
    perror("fstat");
    close(fd);
    return 1;
  }
  size_t len = (size_t)st.st_size;
  if (len == 0) {
    close(fd);
    return 1;
  }
  void *p = mmap(NULL, len, PROT_READ, MAP_PRIVATE, fd, 0);
  if (p == MAP_FAILED) {
    perror("mmap");
    close(fd);
    return 1;
  }
  close(fd);
  int32_t sum = 0;
  const uint8_t *bytes = (const uint8_t *)p;
  for (size_t i = 0; i < len; i++) {
    sum += (int32_t)bytes[i];
  }
  if (munmap(p, len) != 0) {
    perror("munmap");
    return 2;
  }
  return sum & 255;
}

File: tests/bench/io_mmap_throughput.zig
// io_mmap_throughput.zig — I/O 基线：大文件 mmap 顺序扫描（Zig -O2 对照）
const std = @import("std");

pub fn main() !void {
    const path = "tests/bench/.io_mmap_bench_tmp";
    const file = try std.fs.cwd().openFile(path, .{});
    defer file.close();
    const len = try file.getEndPos();
    if (len == 0) std.process.exit(1);
    const map = try std.os.mmap(null, len, std.os.PROT.READ, std.os.MAP.PRIVATE, file.handle, 0);
    defer std.os.munmap(map);
    var sum: i32 = 0;
    for (map) |b| {
        sum +%= @as(i32, @intCast(b));
    }
    std.process.exit(@intCast(sum & 255));
}
